crates/amaru-ledger/src/rules/transaction/metadata.rs

```diff
@@ -0,0 +1,53 @@
+use amaru_kernel::{AuxiliaryData, Bytes, ComputeHash, Hash, TransactionBody};
+
+use crate::rules::TransactionRuleViolation;
+
+pub enum InvalidTransactionMetadata {
+    MissingTransactionMetadata(Bytes),
+    MissingTransactionAuxiliaryDataHash(Hash<32>),
+    ConflictingMetadataHash {
+        supplied: Vec<u8>,
+        expected: Vec<u8>,
+    },
+}
+
+impl Into<TransactionRuleViolation> for InvalidTransactionMetadata {
+    fn into(self) -> TransactionRuleViolation {
+        TransactionRuleViolation::InvalidTransactionMetadata(self)
+    }
+}
+
+// impl From<InvalidTransactionMetadata> for TransactionRuleViolation {
+//     fn from(value: InvalidTransactionMetadata) -> Self {
+//         TransactionRuleViolation::InvalidTransactionMetadata(value)
+//     }
+// }
+
+pub fn validate_metadata(
+    transaction: &TransactionBody,
+    auxilary_data: Option<AuxiliaryData>,
+) -> Result<(), InvalidTransactionMetadata> {
+    match (transaction.auxiliary_data_hash.clone(), auxilary_data) {
```

Comment:
Same here, the clone should be avoided. Introducing lifetime to all errors is likely the way to go.